src/libaktualizr-posix/ipuptanesecondary.cc

```diff
@@ -163,37 +220,136 @@ data::ResultCode::Numeric IpUptaneSecondary::install(const std::string& target_n
   return static_cast<data::ResultCode::Numeric>(r->result);
 }
 
-Manifest IpUptaneSecondary::getManifest() const {
-  LOG_DEBUG << "Getting the manifest from secondary with serial " << getSerial();
+data::ResultCode::Numeric IpUptaneSecondary::install_v2(const Uptane::Target& target) {
+  data::ResultCode::Numeric data_delivery_result = data::ResultCode::Numeric::kDownloadFailed;
+
+  LOG_INFO << "Invoking an image delivery of the target  " << target.filename() << " to/on the Secondary "
+           << getSerial();
+
+  if (target.IsOstree()) {
+    data_delivery_result = downloadOstreeRev(target);
+  } else {
+    data_delivery_result = uploadFirmware(target);
+  }
+
+  if (data::ResultCode::Numeric::kOk != data_delivery_result) {
+    return data_delivery_result;
+  }
+
+  LOG_INFO << "Invoking an installation of the target " << target.filename() << " on the Secondary " << getSerial();
+
+  return invokeInstallOnSecondary(target);
+}
+
+bool IpUptaneSecondary::sendFirmware(const std::string& data) {
+  LOG_INFO << "Sending firmware to the secondary, size: " << data.size();
   Asn1Message::Ptr req(Asn1Message::Empty());
+  req->present(AKIpUptaneMes_PR_sendFirmwareReq);
 
-  req->present(AKIpUptaneMes_PR_manifestReq);
+  auto m = req->sendFirmwareReq();
+  SetString(&m->firmware, data);
+  auto resp = Asn1Rpc(req, getAddr());
+
+  if (resp->present() != AKIpUptaneMes_PR_sendFirmwareResp) {
+    LOG_ERROR << "Failed to get response to sending firmware to Secondary";
+    return false;
+  }
+
+  auto r = resp->sendFirmwareResp();
+  return r->result == AKInstallationResult_success;
+}
+
+data::ResultCode::Numeric IpUptaneSecondary::downloadOstreeRev(const Uptane::Target& target) {
+  LOG_INFO << "Invoking donwload of the ostree target revision ( " << target.sha256Hash() << " ) on Secondary ( "
```

Comment:
Perhaps "Instructing Secondary <xyz> to download OSTree commit <hash>."

Reply:
OK, although IMHO, 'invoke' better reflects what actually happens than 'instruct' as the later implies that Secondary still has freedom not to execute the instruction :)

Reply:
"Invoke" is a slightly more accurate verb but it makes the sentence structure somewhat more complicated, and I think the gain in simplicity is worth it.